docs/rfc/rfc-020-block-structure-consolidation.md

```diff
@@ -0,0 +1,364 @@
+store only some pieces of each block and still achieve a space savings without having
+to change the block structure at all.
+
+The main way to achieve this would be by _no longer saving data that does not change_.
+At each height we save a set of data that is unlikely to have changed from the
+previous height in the block structure, this includes the `ValidatorAddress`es,
+the `ValidatorsHash`, the `ChainID`. These do not need to be saved along with
+_each_ block. We could easily save the value and the height at which the value
+was updated and construct each block using the data that existed at the time.
+
+This document does not make any specific recommendations around storage since
+that is likely to change with upcoming improvements to to the database infrastructure.
+However, it's important to note that removing fields from the block for the
+purposes of 'saving space' may not be that meaningful. We should instead focus
+our attention of removing fields from the block that are no longer needed
+for correct functioning of the protocol.
+
+#### Updates to propagation
+
+Block propagation suffers from the same issue that plagues block storage, we
+propagate all of the contents of each block proto _even when these contents are redundant
+or unchanged from previous blocks_. For example, we propagate the `ValidatorAddress`es
+for each block in the `CommitSig` structure even when it never changed from a
+previous height. We could achieve a speed-up in many cases by communicating the
+hashes _first_ and letting peers request additional information when they do not
+recognize the communicated hash. 
+
+For example, in the case of the `ValidatorAddress`es, the node would first
+communicate the `ValidatorsHash` of the block to its peers. The peers would
+check their storage for a validator set matching the provided hash. If the peer
+has a matching set, it would populate its local block structure with the
+appropriate values from its store. If peer did not have a matching set, it would
+issue a request to its peers, either via P2P or RPC for the data it did not have.
+
+Conceptually, this is very similar to how content addressing works in protocols
+such as git where pushing a commit does not require pushing the entire contents
+of the tree referenced by the commit.
+
+### Impact on light clients
+
+As outlined in the section [On Tendermint Blocks](#on-tendermint-blocks), there
+is a distinction between what data is referenced in the Merkle root hash and the
+contents of the proto structure we currently call the `Block`.
+
+Any changes to the Merkle root hash will necessarily be breaking for legacy light clients.
+Either a soft-upgrades scheme will need to be implemented or a hard fork will
+be required for chains and light clients to function with the new hashes.
+This means that all of the additions and deletions from the Merkle root hash
+proposed by this document will be light client breaking.
```

Comment:
I think the actions proposed in this document would be an excellent first use-case for soft-upgrades, because we don't change any functionality, but only the way fields are organized, hashes are calculated and blocks are verified.